Crashlytics/Crashlytics/Models/Record/FIRCLSReportAdapter.m

```diff
@@ -388,13 +449,66 @@ - (google_crashlytics_Session_Event_Application)protoEventApplication {
   app.background = [self wasInBackground];
   app.ui_orientation = [self uiOrientation];
 
-  // TODO: Where do these come from?
-  //    app.custom_attributes
-  //    app.custom_attributes_count
+  app.custom_attributes = [self protoCustomAttributesWithError:nil];
+  app.custom_attributes_count = (pb_size_t)self.userKeyValues.count;
 
   return app;
 }
 
+- (google_crashlytics_Session_Event_Application)protoEventApplicationForError:
+    (FIRCLSRecordError *)error {
+  google_crashlytics_Session_Event_Application app =
+      google_crashlytics_Session_Event_Application_init_default;
+
+  // TODO: Filter by binaries by stacktrace, Protobuf.scala:93
+  app.execution.binaries = [self protoBinaryImages];
+  app.execution.binaries_count = (pb_size_t)self.binaryImages.count;
+
+  google_crashlytics_Session_Event_Application_Execution_Signal emptySignal =
+      google_crashlytics_Session_Event_Application_Execution_Signal_init_default;
+  app.execution.signal = emptySignal;
+
+  // Create single thread from stacktrace
+  google_crashlytics_Session_Event_Application_Execution_Thread *threads =
+      malloc(sizeof(google_crashlytics_Session_Event_Application_Execution_Thread) * 1);
+  google_crashlytics_Session_Event_Application_Execution_Thread thread =
+      google_crashlytics_Session_Event_Application_Execution_Thread_init_default;
+  thread.frames = [self protoFramesWithStacktrace:error.stacktrace];
+  thread.frames_count = (pb_size_t)error.stacktrace.count;
+  threads[0] = thread;
+  app.execution.threads = threads;
+  app.execution.threads_count = 1;
+
+  app.background = [self wasInBackground];
+  app.ui_orientation = [self uiOrientation];
+
+  app.custom_attributes = [self protoCustomAttributesWithError:error];
+  app.custom_attributes_count = (pb_size_t)self.userKeyValues.count + 2;
+
+  return app;
+}
+
+/// Generate an array of CustomAttributes from the user defined key values.
+/// For recorded errors, the error's nserror-domain and nserror-code are also added.
+/// @param error Error from recordError API
+- (google_crashlytics_CustomAttribute *)protoCustomAttributesWithError:(FIRCLSRecordError *)error {
+  NSDictionary<NSString *, NSString *> *kvs = [self KeyValuesWithError:error];
+  NSUInteger size = kvs.count;
+
+  google_crashlytics_CustomAttribute *attributes =
+      malloc(sizeof(google_crashlytics_CustomAttribute) * size);
```

Comment:
If size is zero, might want to return? Just a suggestion but leave it up to you. Not sure what happens with nanopb in the case where it gets a zero-length, non NULL pointer (if that happens to be the semantics of malloc(0) on iOS).

https://stackoverflow.com/questions/2022335/whats-the-point-in-malloc0

Reply:
I just verified that when size = 0, nanopb treats the fields as not set and that I am able to still parse the nanopb bytes.